File: Semaphore_java/src/Producer.java
public class Producer extends Thread {
    private final Integer _itemsCount;
    private final Integer _pushDelay_ms;
    private final Storage _storage;

    public Producer(int itemsCount, int pushDelay_ms, Storage storage) {
        _itemsCount = itemsCount;
        _pushDelay_ms = pushDelay_ms;
        _storage = storage;
    }

    @Override
    public void run(){
        for (Integer i = 0; i < _itemsCount; i++)
        {
            try {
                _storage.PushItemSemaphore.acquire();
                Thread.sleep(_pushDelay_ms);
                _storage.AccessToStorageSemaphore.acquire();

                Main.CounterAccess.acquire();
                String item = "item " + Main.Counter++;
                Main.CounterAccess.release();
                _storage.pushItem(item);
                System.out.println("Push " + item);

                _storage.AccessToStorageSemaphore.release();
                _storage.PickItemSemaphore.release();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }
}

File: Semaphore_sharp/Semaphore_sharp/Producer.cs
using System;
using System.Threading;

namespace Semaphore_sharp
{
    public class Producer
    {
        private int _itemsCount;
        private int _pushDelay_ms;
        private Storage _storage;

        public Producer(int itemsCount, int pushDelay_ms, Storage storage)
        {
            _itemsCount = itemsCount;
            _pushDelay_ms = pushDelay_ms;
            _storage = storage;
        }

        public void Start()
        {
            for (int i = 0; i < _itemsCount; i++)
            {
                _storage.PushItemSemaphore.WaitOne();
                Thread.Sleep(_pushDelay_ms);
                _storage.AccessToStorageSemaphore.WaitOne();

                Program.CounterAccess.WaitOne();
                string item = $"Item {Program.ItemsCounter++}";
                Program.CounterAccess.Release();
                _storage.PushItem(item);

                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Push {item}");
                Console.ForegroundColor = ConsoleColor.Black;

                _storage.AccessToStorageSemaphore.Release();
                _storage.PickItemSemaphore.Release();
            }
        }
    }
}